docs/proposal/20211210-launch-templates-managedmachinepools.md

```diff
@@ -0,0 +1,254 @@
+
+This proposal outlines changes to CAPA that will introduce new capabilities to utilise launch templates for `AWSManagedMachinePool` and brings its functionality in line with `AWSMachinePool`.
+
+## Motivation
+
+We are increasingly hearing requests from users of CAPA that a particular feature / configuration option isn't exposed by CAPAs implementation of managed machine pools (i.e. `AWSManagedMachinePool`) and on investigation the feature is available via a launch template (nitro enclaves or placement as an example). It some instances, users of CAPA have had to use unmanaged machine pools (i.e. `AWSMachinePool`) instead.
+
+The motivation is to improve consistency between the2 varieties of machine pools and expose to the user features of launch templates.
+
+> Note: it may not be completely consistent in the initial implementation as we may need to deprecate some API definitions over time but the plan will be to be eventually consistent ;)
+
+### Goals
+
+* Consistent API to declare a use launch templates for `AWSMachinePool` and `AWSManagedMachinePool`
+* Single point of reconciliation of launch templates
+* Guide to the deprecation of certain API elements in `AWSManagedMachinePool`
+
+### Non-Goals/Future Work
+
+* Add non-existent controller unit tests for `AWSMachinePool` and `AWSManagedMachinePool`
+
+## Proposal
+
+At a high level, the plan is to:
+
+1. Add a new `AWSLaunchTemplate` field to [AWSManagedMachinePoolSpec](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/exp/api/v1beta1/awsmanagedmachinepool_types.go#L65) that uses the existing [AWSLaunchTemplate](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/ec057ad6e613a6578f67bf68a6c77fbe772af933/exp/api/v1beta1/types.go#L58) struct. For example:
+
+```go
+// AWSLaunchTemplate specifies the launch template and version to use when an instance is launched. This field 
+// will become mandatory in the future and its recommended you use this over fields AMIType,AMIVersion,InstanceType,DiskSize,InstanceProfile.
+// +optional
+AWSLaunchTemplate AWSLaunchTemplate `json:"awsLaunchTemplate"`
+```
+
+2. Update the comments on the below fields of [AWSManagedMachinePoolSpec](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/9bc29570614aa7123d79f042b6e6efc2aaf3e490/exp/api/v1beta1/awsmanagedmachinepool_types.go#L65) to indicate that the fields is deprecated and that `AWSlaunchTemplate` should be used.
```

Comment:
Thanks @richardcase for writing this detailed proposal.
I have few doubts for my clarifications: If a user is using older version of CAPA where AWSLaunch template in Managed Machinepool is not supported and they want to upgrade then how will we handle the conversion ? How will we get the launch template ID and other related info?

Reply:
As we need to support the deprecated fields for a number of versions, i propose that we automatically create the `AWSLaunchTemplate` on conversion when we bump the API version. This would be handled in the conversion functions.